### ISSUE TITLE
http.c 2146 have a bug

### ISSUE THREAD

**Issue by jadesnake** [curl/curl#782]
http.c at 2139 line data->state.resume_from if i do not set CURLOPT_RESUME_FROM and CURLOPT_SEEKFUNCTION , the CURLOPT_READFUNCTION do not work.

httpc. at 2191 line while(passed < data->state.resume_from);   resume_from is stream start point and  passed is read point ,then the code while read only 1 byte. so the update stream can not read complete.

**Comment by bagder** [comment]
Can you please provide an example of what you do when this cause a problem so that we can reproduce it?

**Comment by jadesnake** [comment]
size_t StreamUpdate(void *ptr, size_t size, size_t nmemb, void *userdata)
	{
		std::stringstream *buf = static_cast<std::stringstream*>(userdata);
		std::string test = buf->str();
		buf->read((char*)ptr,size*nmemb);
		return buf->gcount();
	}
	void	CHttpClient::PerformParam(const CAtlString& url)
	{
		curl_easy_reset(m_url);
		m_wbuf.clear();	
		m_wbuf.str("");
		m_wbuf.flush();
		m_rbuf.clear();	
		m_headbuf.clear();
		m_headbuf.str("");
		{
			if( m_headerlist )
			{
				curl_slist_free_all(m_headerlist);
				m_headerlist = NULL;
			}
			mapStrings::iterator itPos = m_header.begin();
			for(itPos;itPos!=m_header.end();itPos++)
			{
				CAtlString strVal;
				strVal.Format(_T("%s:%s\r\n"),itPos->first,itPos->second);
				m_headerlist = curl_slist_append(m_headerlist,(char*)CT2CA(strVal,CP_UTF8));
			}
		}
		{
			CAtlStringA strTmp = (char*)CT2CA(url);
			
			curl_easy_setopt(m_url, CURLOPT_URL, strTmp.GetString() );
		}
		curl_easy_setopt(m_url, CURLOPT_CONNECTTIMEOUT,10);		
		curl_easy_setopt(m_url, CURLOPT_HTTPHEADER, m_headerlist);
		curl_easy_setopt(m_url, CURLOPT_USERAGENT,(char*)CT2CA(m_agent,CP_UTF8));
		curl_easy_setopt(m_url, CURLOPT_TIMEOUT,m_tmOut/1000);
		curl_easy_setopt(m_url, CURLOPT_WRITEFUNCTION,StreamSave);
		curl_easy_setopt(m_url, CURLOPT_WRITEDATA,(void*)&m_wbuf);
		if (m_rbuf.tellp()) {
			curl_off_t nLen = (curl_off_t)m_rbuf.tellp();
			curl_easy_setopt(m_url, CURLOPT_READFUNCTION, StreamUpdate);
			curl_easy_setopt(m_url, CURLOPT_READDATA, (void*)&m_rbuf);
			curl_easy_setopt(m_url, CURLOPT_INFILESIZE, nLen);
			curl_easy_setopt(m_url, CURLOPT_INFILESIZE_LARGE, nLen);
			curl_easy_setopt(m_url, CURLOPT_PUT,1L);
			curl_easy_setopt(m_url, CURLOPT_UPLOAD, 1L);
		}
		if( m_tgProxy.nType!=Proxy::NONE )
		{
			curl_easy_setopt(m_url, CURLOPT_PROXYTYPE,m_tgProxy.nType);
			curl_easy_setopt(m_url, CURLOPT_PROXYPORT,m_tgProxy.nPort);
			curl_easy_setopt(m_url, CURLOPT_PROXY_SERVICE_NAME,(char*)CT2CA(m_tgProxy.strServer) );
			curl_easy_setopt(m_url, CURLOPT_PROXYUSERNAME,(char*)CT2CA(m_tgProxy.strName));
			curl_easy_setopt(m_url, CURLOPT_PROXYUSERPWD,(char*)CT2CA(m_tgProxy.strPass));
		}
		if( m_postparam )
		{
			curl_easy_setopt(m_url, CURLOPT_HTTPPOST,m_postparam);
		}
		if (m_bWriteHeader) {
			curl_easy_setopt(m_url, CURLOPT_HEADERDATA, (void*)&m_headbuf);
			curl_easy_setopt(m_url, CURLOPT_HEADERFUNCTION, StreamHeader);
		}
		curl_easy_setopt(m_url, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
		#if defined(_DEBUG)
			curl_easy_setopt(m_url, CURLOPT_VERBOSE, 1L);
			curl_easy_setopt(m_url, CURLOPT_DEBUGFUNCTION, dbg_trace);
		#endif
		m_header.clear();
	}
       curl_easy_perform(url);

**Comment by jadesnake** [comment]
when i use curl_formadd , the formdata do not send.when send it ?
http.c code run here
 if(result)
      failf(data, "Failed sending POST request");
    else
      /* setup variables for the upcoming transfer */
      Curl_setup_transfer(conn, FIRSTSOCKET, -1, TRUE,
                          &http->readbytecount, FIRSTSOCKET,
                          &http->writebytecount);

    if(result) {
      Curl_formclean(&http->sendit); /* free that whole lot */
      return result;
    }
    /* convert the form data */
    result = Curl_convert_form(data, http->sendit);
    if(result) {
      Curl_formclean(&http->sendit); /* free that whole lot */
      return result;
    }
but http->sendit keep value.
debug trace : 
=> Send headerPOST /businessservlet HTTP/1.1
Host: 192.168.70.181:7002
Accept: */*
ACTIONCODE:5002

CREATE_DATE:2016/50/04/29/16

HAS_BODY:1

IS_COMPRESS:0

IS_CRYPT:0

JKVERSION:V1.0.055

RESERVE_HEAD:tdp

SENDER:SERVYOU001

SWJG_DM:19901000000

Content-Length: 491
Expect: 100-continue
Content-Type: multipart/form-data; boundary=------------------------f14bf03d3e7d425a


<= Recv headerHTTP/1.1 404 Not Found

<= Recv headerDate: Fri, 29 Apr 2016 01:08:29 GMT

<= Recv headerContent-Length: 1164

<= Recv headerContent-Type: text/html; charset=UTF-8

<= Recv headerX-Powered-By: Servlet/2.5 JSP/2.1

<= Recv header

<= Recv data<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Draft//EN">
<HTML>
<HEAD>
<TITLE>Error 404--Not Found</TITLE>
</HEAD>
<BODY bgcolor="white">
<FONT FACE=Helvetica><BR CLEAR=all>
<TABLE border=0 cellspacing=5><TR><TD><BR CLEAR=all>
<FONT FACE="Helvetica" COLOR="black" SIZE="3"><H2>Error 404--Not Found</H2>
</FONT></TD></TR>
</TABLE>
<TABLE border=0 width=100% cellpadding=10><TR><TD VALIGN=top WIDTH=100% BGCOLOR=white><FONT FACE="Courier New"><FONT FACE="Helvetica" SIZE="3"><H3>From RFC 2068 <i>Hypertext Transfer Protocol -- HTTP/1.1</i>:</H3>
</FONT><FONT FACE="Helvetica" SIZE="3"><H4>10.4.5 404 Not Found</H4>
</FONT><P><FONT FACE="Courier New">The server has not found anything matching the Request-URI. No indication is given of whether the condition is temporary or permanent.</p><p>If the server does not wish to make this information available to the client, the status code 403 (Forbidden) can be used instead. The 410 (Gone) status code SHOULD be used if the server knows, through some internally configurable mechanism, that an old resource is permanently unavailable and has no forwarding address.</FONT></P>
</FONT></TD></TR>
</TABLE>

</BODY>
</HTML>

**Comment by jadesnake** [comment]
http->sendit is not send ，but clean ? 
if(HTTPREQ_POST_FORM == data->set.httpreq) {
    data->req.bytecount = http->readbytecount + http->writebytecount;

    Curl_formclean(&http->sendit); /* Now free that whole lot */
    if(http->form.fp) {
      /* a file being uploaded was left opened, close it! */
      fclose(http->form.fp);
      http->form.fp = NULL;
    }
  }

**Comment by jadesnake** [comment]
HTTPREQ_POST_FORM  :
...
result = Curl_add_buffer_send(req_buffer, conn, &data->info.request_size, 0, FIRSTSOCKET);
....
formdata is body part,and then  http->writebytecount == 0 ,so the trancsfer do not send body.so the formdata part can not send, but  HTTPREQ_POST is right.

**Comment by jadesnake** [comment]
read bug i fixed code like this.
//fixed by jade for read stream
CURLcode readfun(struct SessionHandle *data, struct connectdata *conn) {
	bool bresume = false;
	int seekerr = CURL_SEEKFUNC_OK;
	if (data->state.resume_from < 0) {
		data->state.resume_from = 0;
	}
	if (data->state.infilesize>0) {
		data->state.infilesize -= data->state.resume_from;
		if (data->state.infilesize <= 0) {
			failf(data, "File already completely uploaded");
			return CURLE_PARTIAL_FILE;
		}
	}
	else {
		//缺少文件长度
		return CURLE_OK;
	}
	if (conn->seek_func && !data->state.this_is_a_follow) {
		seekerr = conn->seek_func(conn->seek_client, data->state.resume_from, SEEK_SET);
	}
	if (seekerr != CURL_SEEKFUNC_OK) {
		if (seekerr != CURL_SEEKFUNC_CANTSEEK) {
			failf(data, "Could not seek stream");
			return CURLE_READ_ERROR;
		}
	}
	curl_off_t passed = 0;
	do {
		size_t readthisamountnow = BUFSIZE;
		if (data->state.resume_from - passed > CURL_OFF_T_C(BUFSIZE))
			readthisamountnow = BUFSIZE;
		else if(data->state.resume_from)
			curlx_sotouz(data->state.resume_from - passed);
		size_t actuallyread =
			data->state.fread_func(data->state.buffer, 1, readthisamountnow,
				data->state.in); //why read to buffer,it used dowload 
		passed += actuallyread;
		if ((actuallyread == 0) || (actuallyread > readthisamountnow)) {
			failf(data, "Could only read %" CURL_FORMAT_CURL_OFF_T
				" bytes from the input", passed);
			return CURLE_READ_ERROR;
		}
	} while (passed < data->state.infilesize);
	return CURLE_OK;
}

**Comment by jay** [comment]
Please wrap your code in markdown by starting and ending it with `~~~`. You can also add a 'c' after the first one to indicate the code is in the C language. I've modified all your previous posts to do that. For example:

`~~~c`
`if(code) etc`
`~~~`

`~~~`
`other output`
`~~~`

That said this thread is really hard to follow. You did not provide your libcurl version information, and I'm having trouble understanding you. The best thing you can give us is a self-contained example with the minimum amount of code needed to reproduce. I don't know where readfun came from, I can't find it in the repo history.

**Comment by jadesnake** [comment]
size_t StreamUpdate(void *ptr, size_t size, size_t nmemb, void *userdata)
{
	std::stringstream *buf = static_cast<std::stringstream*>(userdata);
	std::string test = buf->str();
	buf->read((char*)ptr,size*nmemb);
	return buf->gcount();
}
int main(int argc, char *argv[])
{
	std::stringstream	m_rbuf;

	CURL		*m_url = curl_easy_init();;
	curl_easy_setopt(m_url, CURLOPT_CONNECTTIMEOUT,10);
	curl_easy_setopt(m_url, CURLOPT_USERAGENT,"Mozilla/3.0 (compatible; Indy Library)");
	curl_easy_setopt(m_url, CURLOPT_TIMEOUT,1000);	
	curl_easy_setopt(m_url, CURLOPT_WRITEFUNCTION,StreamSave);
	curl_easy_setopt(m_url, CURLOPT_WRITEDATA,(void*)&m_wbuf);

	m_rbuf << "i am a simple data.";

	curl_off_t nLen = (curl_off_t)m_rbuf.tellp();
	curl_easy_setopt(m_url, CURLOPT_READFUNCTION, StreamUpdate);
	curl_easy_setopt(m_url, CURLOPT_READDATA, (void*)&m_rbuf);
	curl_easy_setopt(m_url, CURLOPT_INFILESIZE, nLen);
	curl_easy_setopt(m_url, CURLOPT_INFILESIZE_LARGE, nLen);
	curl_easy_setopt(m_url, CURLOPT_PUT,1L);
	curl_easy_setopt(m_url, CURLOPT_UPLOAD, 1L);
	curl_easy_setopt(m_url, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
	curl_easy_perform(m_url);
	return 0;
}
version 7.48
http.c have data->state.resume_from . thanks your response.

**Comment by jadesnake** [comment]
I found my bug ，custom header have more "\r\n" , thanks.
